text/0000-closure-functions.md

```diff
@@ -0,0 +1,171 @@
+
+```rust
+let mut calls = 0;
+
+// If Neon allowed `FnMut`, this could be written:
+let f = JsFunction::new(move |mut cx| {
+    let current = calls;
+    let f = cx.argument::<JsFunction>(0)?;
+    let this = cx.undefined();
+
+    // If `f` is a reference to this `JsFunction`, it will be called
+    // recursively; each time incrementing `calls`. However, this is
+    // unsound because the currently running code should have exclusive
+    // ownership.
+    f.call(&mut cx, this, [f])?
+
+    calls = current + 1;
+
+    Ok(cx.number(calls))
+});
+```
+
+Unlike closures used in threading contexts (e.g., `Channel`, `TaskBuilder`), closures wrapped in a `JsFunction` do not need to implement `Send`. This is because they are only ever called from the event loop executing on a single thread.
+
+Since the closures do not need to be `Send`, `Rc<RefCell<T>>` is a useful pattern for sharing state between methods.
+
+```rust
+#[derive(Clone, Default)]
+struct Stats {
+    calls: Rc<RefCell<usize>>,
+}
+
+impl Stats {
+    fn inc(&self) {
+        *self.calls.borrow_mut() += 1;
+    }
+
+    fn calls(&self) -> usize {
+        self.calls.borrow()
+    }
+}
+
+#[neon::main]
+fn main(mut cx: ModuleContext) -> NeonResult<()> {
+    let stats = Stats::default();
+
+    cx.export_function("get_a", {
+        let stats = stats.clone();
+
+        move |mut cx| {
+            stats.inc();
+            Ok(cx.string("A"))
+        }
+    });
+
+    cx.export_function("get_b", {
+        let stats = stats.clone();
+
+        move |mut cx| {
+            stats.inc();
+            Ok(cx.string("B"))
+        }
+    });
+
+    cx.export_function("calls", {
+        let stats = stats.clone();
+
+        move |mut cx| Ok(cx.number(stats.calls()))
+    });
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Calling functions could be slower by a very small amount from needing to perform dynamic dispatch.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+The primary alternative considered was to allow closures on *all* Node-API versions. However, it is very subtle that this would leak. Especially since the leak is determined by Neon feature flags and not based on the current runtime.
+
+Instead, we will only provide this feature for Node-API 5+. This costs a small amount of code complexity; however, we may be able to remove this later when all versions of Node on our support matrix provide Node-API 5.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* ~~Do closures need to be `Send`?~~ `V8` heavily uses thread local storage for the vent loop. It would be an extremely large change to be able to switch the thread running the event loop. It would also require changes to `libuv`. Even thought it's not _guaranteed_ this won't change, it seems unlikely enough that the small risk is worth the improved API.
```

Comment:
```suggestion
* ~~Do closures need to be `Send`?~~ `V8` heavily uses thread local storage for the event loop. It would be an extremely large change to be able to switch the thread running the event loop. It would also require changes to `libuv`. Even thought it's not _guaranteed_ this won't change, it seems unlikely enough that the small risk is worth the improved API.
```